pallets/system/src/data_root_builder.rs

```diff
@@ -0,0 +1,355 @@
+use codec::Decode;
+use da_primitives::asdr::AppExtrinsic;
+use frame_support::traits::ExtrinsicCall;
+use sp_core::H256;
+use sp_runtime::traits::Extrinsic;
+use sp_std::vec::Vec;
+
+const LOG_TARGET: &str = "runtime::system::data_root_builder";
+
+pub type DRFOutput = Option<Vec<u8>>;
+pub type DRFCallOf<T> = <T as Extrinsic>::Call;
+
+pub trait DataRootFilter {
+	type UncheckedExtrinsic: Decode + ExtrinsicCall;
+
+	fn filter(call: &DRFCallOf<Self::UncheckedExtrinsic>) -> DRFOutput;
+}
+
+pub trait DataRootBuilder<F: DataRootFilter> {
+	fn build<'a, I>(app_extrinsics: I) -> H256
+	where
+		I: IntoIterator<Item = &'a AppExtrinsic>,
+	{
+		let mut used_extrinsics_count = 0u32;
+		let mut total_extrinsics_count = 0u32;
+
+		let filtered_iter = app_extrinsics
+			.into_iter()
+			.enumerate()
+			.filter_map(|(idx, app_extrinsic)| {
+				// Decode call and log any failure.
+				total_extrinsics_count += 1;
+				match F::UncheckedExtrinsic::decode(&mut app_extrinsic.data.as_slice()) {
+					Ok(app_unchecked_extrinsic) => {
+						// Some((app_unchecked_extrinsic, app_extrinsic.data.as_slice()))
+						Some(app_unchecked_extrinsic)
+					},
+					Err(err) => {
+						log::error!(
```

Comment:
This will log error and skip invalid extrinsic? Should case of badly encoded extrinsic stop node and demand bug fix, or its ok to ignore it (and monitor logs)?